File: java/src/main/java/com/alxarafe/app/application/catalogue/CreateItem.java
package com.alxarafe.app.application.catalogue;

import com.alxarafe.app.domain.catalogue.entity.Item;
import com.alxarafe.app.domain.catalogue.valueobject.ItemFamilyCode;
import com.alxarafe.app.domain.catalogue.valueobject.ItemId;
import com.alxarafe.app.domain.catalogue.valueobject.Money;
import com.alxarafe.app.domain.catalogue.valueobject.Sku;
import com.alxarafe.app.domain.catalogue.valueobject.UomCode;

public final class CreateItem {
    private final ItemRepository items;
    private final ItemFamilyRepository families;
    private final UomRepository uoms;

    public CreateItem(ItemRepository items, ItemFamilyRepository families, UomRepository uoms) {
        this.items = items;
        this.families = families;
        this.uoms = uoms;
    }

    public Item execute(String sku, String name, String familyCode, String baseUomCode,
                        boolean isBatchManaged, boolean isExpirable,
                        double baseCost, String currency) {
        if (name.length() > 255) {
            throw new IllegalArgumentException("Item name cannot exceed 255 characters.");
        }

        var family = families.findByCode(new ItemFamilyCode(familyCode))
                .orElseThrow(() -> new IllegalArgumentException("Item family code does not exist."));

        var uom = uoms.findByCode(new UomCode(baseUomCode))
                .orElseThrow(() -> new IllegalArgumentException("Base unit of measure code does not exist."));

        Item item = new Item(ItemId.generate(), new Sku(sku), name, family.id(), uom.id(),
                isBatchManaged, isExpirable, new Money(baseCost, currency));

        if (items.findBySku(item.sku()).isPresent()) {
            throw new ItemConflict("Item sku already exists.");
        }

        items.save(item);
        return item;
    }
}

File: php/src/Application/Catalogue/CreateItem.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Application\Catalogue;

use Alxarafe\App\Domain\Catalogue\Entity\Item;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemFamilyCode;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemId;
use Alxarafe\App\Domain\Catalogue\ValueObject\Sku;
use Alxarafe\App\Domain\Catalogue\ValueObject\UomCode;
use InvalidArgumentException;

final readonly class CreateItem
{
    public function __construct(
        private ItemRepository $items,
        private ItemFamilyRepository $families,
        private UomRepository $uoms,
    ) {
    }

    public function execute(
        string $sku,
        string $name,
        string $familyCode,
        string $baseUomCode,
        bool $isBatchManaged,
        bool $isExpirable,
    ): Item {
        if (strlen($name) > 255) {
            throw new InvalidArgumentException('Item name cannot exceed 255 characters.');
        }

        $family = $this->families->findByCode(new ItemFamilyCode($familyCode));
        if ($family === null) {
            throw new InvalidArgumentException('Item family code does not exist.');
        }

        $uom = $this->uoms->findByCode(new UomCode($baseUomCode));
        if ($uom === null) {
            throw new InvalidArgumentException('Base unit of measure code does not exist.');
        }

        $item = new Item(
            ItemId::generate(),
            new Sku($sku),
            $name,
            $family->id(),
            $uom->id(),
            $isBatchManaged,
            $isExpirable,
        );

        if ($this->items->findBySku($item->sku()) !== null) {
            throw new ItemConflict('Item sku already exists.');
        }

        $this->items->save($item);
        return $item;
    }
}
